src/libfsm/print/cdata.c

```diff
@@ -0,0 +1,1234 @@
+save_state_eager_outputs(struct cdata_config *config, const struct ir_state_eager_output *eager_outputs, size_t *offset)
+{
+	if (eager_outputs == NULL || eager_outputs->count == 0) {
+		assert(*offset == STATE_OFFSET_NONE);
+		return true;
+	}
+
+	/* These must be in ascending order. */
+	for (size_t i = 1; i < eager_outputs->count; i++) {
+		assert(eager_outputs->ids[i - 1] < eager_outputs->ids[i]);
+	}
+
+#if REUSE_EAGER_OUTPUT_SETS
+#if REUSE_NAIVE || EXPENSIVE_CHECKS
+	/* Linear scan. See comments about reuse in save_state_endids. */
+	size_t naive_offset = STATE_OFFSET_NONE;
+	for (size_t b_i = 0; b_i < config->eager_output_buf.used; b_i++) {
+		size_t e_i;
+		for (e_i = 0; e_i < eager_outputs->count; e_i++) {
+			if (b_i + e_i >= config->eager_output_buf.used) {
+				break; /* reached the end, not found */
+			}
+			if (config->eager_output_buf.buf[b_i + e_i] != eager_outputs->ids[e_i]) {
+				break; /* mismatch */
+			}
+		}
+
+		/* If there's a potential match, check that it isn't followed by
+		 * a value > the last, because it would falsely continue the run. */
+		if (e_i == eager_outputs->count
+		    && b_i + e_i + 1 < config->eager_output_buf.used
+		    && config->eager_output_buf.buf[b_i + e_i + 1] <= eager_outputs->ids[e_i - 1]) {
+			naive_offset = b_i;
+			break;
+		}
+	}
+
+	if (REUSE_NAIVE) {
+		*offset = naive_offset;
+	}
+#endif	/* REUSE_NAIVE || EXPENSIVE_CHECKS */
+
+	/* TODO: better impl */
+
+#if LOG_REUSE
+	if (*offset == STATE_OFFSET_NONE) {
+		config->reuse_stats_eager_output.miss++;
+	} else {
+		config->reuse_stats_eager_output.hit++;
+	}
+#endif
+
+	if (*offset != STATE_OFFSET_NONE) {
+		return true;
+	}
+#endif	/* REUSE_EAGER_OUTPUT_SETS */
+
+	/* If necessary add a 0, as in save_state_endids above. */
+	if (config->eager_output_buf.used > 0
+	    && eager_outputs->ids[0] > config->eager_output_buf.buf[config->eager_output_buf.used - 1]) {
+		if (!append_eager_output(&config->eager_output_buf, 0)) {
+			return false;
+		}
+	}
+
+	const size_t base = config->eager_output_buf.used;
+
+	for (size_t i = 0; i < eager_outputs->count; i++) {
+		if (eager_outputs->ids[i] > config->max_eager_output_id) {
+			config->max_eager_output_id = eager_outputs->ids[i];
+		}
+
+		if (!append_eager_output(&config->eager_output_buf, eager_outputs->ids[i])) {
+			return false;
+		}
+	}
+
+	assert(base != STATE_OFFSET_NONE);
+	*offset = base;
+	return true;
+}
+
+struct range_info {
+	uint8_t start;
+	uint8_t end;
+	uint32_t dst_state;
+};
+
+static int
+cmp_outgoing(const void *pa, const void *pb)
+{
+	const struct range_info *a = (const struct range_info *)pa;
+	const struct range_info *b = (const struct range_info *)pb;
+	return a->start < b->start ? -1 : a->start > b->start ? 1 : 0;
+}
+
+static bool
+append_dst(struct dst_buf *buf, uint32_t dst)
+{
+	if (buf->used == buf->ceil) {
+		const size_t nceil = buf->ceil == 0 ? 8 : 2*buf->ceil;
+		uint32_t *nbuf = realloc(buf->buf, nceil * sizeof(nbuf[0]));
+		assert(nbuf != NULL);
```

Comment:
we shouldn't be asserting on realloc in library code. This ought to use the `f_realloc()` interface for user callbacks, and bubble up returning on failure rather than aborting

Reply:
Pass in `fsm->alloc` if you need it

Reply:
None of the existing print interfaces here pass in alloc: https://github.com/katef/libfsm/blob/main/src/libfsm/print.h#L58-L72

Should I add a new typedef?

Reply:
I'm removing the asserting, either way.

Reply:
Removed asserts in 25a9141.

Should I add a new print function typedef that passes in alloc?

Reply:
Not a new one, if you need it here, it's fine to add it for `fsm_print_ir`

Reply:
All right, I'll go that route then.

Reply:
Updated in 4fe0beb8.